<a href="https://colab.research.google.com/github/LokeRuiKee/ml-dl-playground/blob/main/RNN_experiment1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# RNN Template Code

In [ ]:
import numpy as np
import tensorflow as tf
# Define the sequence length and number of features
sequence_length = 10
num_features = 1
# assume each element in the sequence is a single feature
# Generate some sample sequential data
X_train = np.random.randn(100, sequence_length, num_features)
y_train = np.random.randint(0, 2, size=(100,))
# Binary classification task, random labels for demonstration
# randn: NumPy function used to generate an array of random numbers
# from a standard normal distribution (mean = 0, standard deviation = 1)
# randint: NumPy function used to generate random integers
# within a specified range

In [ ]:
# Define the RNN model
model = tf.keras.Sequential([
    tf.keras.layers.SimpleRNN(32, input_shape=(sequence_length, num_features)),
    tf.keras.layers.Dense(1, activation='sigmoid')  # Binary classification output
])
# Compile the model
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
# Train the model
model.fit(X_train, y_train, epochs=10, batch_size=32)
# Once trained, you can use the model for predictions
# For example:
# predictions = model.predict(X_test)


Epoch 1/10
4/4 [==============================] - 1s 7ms/step - loss: 0.7026 - accuracy: 0.5100
Epoch 2/10
4/4 [==============================] - 0s 6ms/step - loss: 0.6835 - accuracy: 0.5500
Epoch 3/10
4/4 [==============================] - 0s 6ms/step - loss: 0.6725 - accuracy: 0.6100
Epoch 4/10
4/4 [==============================] - 0s 6ms/step - loss: 0.6655 - accuracy: 0.6100
Epoch 5/10
4/4 [==============================] - 0s 6ms/step - loss: 0.6592 - accuracy: 0.6100
Epoch 6/10
4/4 [==============================] - 0s 6ms/step - loss: 0.6537 - accuracy: 0.6200
Epoch 7/10
4/4 [==============================] - 0s 6ms/step - loss: 0.6476 - accuracy: 0.6400
Epoch 8/10
4/4 [==============================] - 0s 6ms/step - loss: 0.6407 - accuracy: 0.6300
Epoch 9/10
4/4 [==============================] - 0s 6ms/step - loss: 0.6347 - accuracy: 0.6100
Epoch 10/10
4/4 [==============================] - 0s 5ms/step - loss: 0.6295 - accuracy: 0.6500


# 7 activation, optimizer = adam

## relu

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import SimpleRNN, Dense
from tensorflow.keras.utils import to_categorical

# Step 1: Load and preprocess the MNIST dataset
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train.reshape(-1, 28, 28) / 255.0  # Reshape and normalize training data
x_test = x_test.reshape(-1, 28, 28) / 255.0    # Reshape and normalize test data
y_train = to_categorical(y_train, num_classes=10)  # One-hot encode training labels
y_test = to_categorical(y_test, num_classes=10)    # One-hot encode test labels

# Step 2: Define the RNN model
model = Sequential([
    SimpleRNN(32, input_shape=(28, 28), activation = 'relu'),  # RNN layer with 32 units
    Dense(10, activation='softmax')       # Output layer with 10 units for 10 classes
])

# Step 3: Compile the model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Step 4: Train the model
model.fit(x_train, y_train, epochs=5, batch_size=128, validation_split=0.2)

# Step 5: Evaluate the model on test data
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print('Test Loss: ', test_loss)
print('Test Accuracy: ', test_accuracy)


Epoch 1/5
375/375 [==============================] - 5s 11ms/step - loss: 1.3311 - accuracy: 0.5158 - val_loss: 0.7825 - val_accuracy: 0.7362
Epoch 2/5
375/375 [==============================] - 5s 13ms/step - loss: 0.6660 - accuracy: 0.7854 - val_loss: 0.5357 - val_accuracy: 0.8292
Epoch 3/5
375/375 [==============================] - 4s 10ms/step - loss: 0.5034 - accuracy: 0.8445 - val_loss: 0.4229 - val_accuracy: 0.8764
Epoch 4/5
375/375 [==============================] - 3s 9ms/step - loss: 0.4200 - accuracy: 0.8739 - val_loss: 0.3754 - val_accuracy: 0.8898
Epoch 5/5
313/313 [==============================] - 1s 3ms/step - loss: 0.3319 - accuracy: 0.8966
Test Loss:  0.3319472372531891
Test Accuracy:  0.8966000080108643


## LeakyReLU

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import SimpleRNN, Dense
from tensorflow.keras.utils import to_categorical

# Step 1: Load and preprocess the MNIST dataset
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train.reshape(-1, 28, 28) / 255.0  # Reshape and normalize training data
x_test = x_test.reshape(-1, 28, 28) / 255.0    # Reshape and normalize test data
y_train = to_categorical(y_train, num_classes=10)  # One-hot encode training labels
y_test = to_categorical(y_test, num_classes=10)    # One-hot encode test labels

# Step 2: Define the RNN model
model = Sequential([
    SimpleRNN(32, input_shape=(28, 28), activation = 'LeakyReLU'),  # RNN layer with 32 units
    Dense(10, activation='softmax')       # Output layer with 10 units for 10 classes
])

# Step 3: Compile the model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Step 4: Train the model
model.fit(x_train, y_train, epochs=5, batch_size=128, validation_split=0.2)

# Step 5: Evaluate the model on test data
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print('Test Loss: ', test_loss)
print('Test Accuracy: ', test_accuracy)


Epoch 1/5
375/375 [==============================] - 6s 12ms/step - loss: 1.1511 - accuracy: 0.5963 - val_loss: 0.6466 - val_accuracy: 0.7747
Epoch 2/5
375/375 [==============================] - 4s 10ms/step - loss: 0.6169 - accuracy: 0.7944 - val_loss: 0.5002 - val_accuracy: 0.8398
Epoch 3/5
375/375 [==============================] - 4s 12ms/step - loss: 0.4916 - accuracy: 0.8442 - val_loss: 0.4165 - val_accuracy: 0.8666
Epoch 4/5
375/375 [==============================] - 5s 14ms/step - loss: 0.4188 - accuracy: 0.8695 - val_loss: 0.3602 - val_accuracy: 0.8907
Epoch 5/5
313/313 [==============================] - 1s 5ms/step - loss: 0.3116 - accuracy: 0.9038
Test Loss:  0.3115786612033844
Test Accuracy:  0.9038000106811523


## elu

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import SimpleRNN, Dense
from tensorflow.keras.utils import to_categorical

# Step 1: Load and preprocess the MNIST dataset
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train.reshape(-1, 28, 28) / 255.0  # Reshape and normalize training data
x_test = x_test.reshape(-1, 28, 28) / 255.0    # Reshape and normalize test data
y_train = to_categorical(y_train, num_classes=10)  # One-hot encode training labels
y_test = to_categorical(y_test, num_classes=10)    # One-hot encode test labels

# Step 2: Define the RNN model
model = Sequential([
    SimpleRNN(32, input_shape=(28, 28), activation = 'elu'),  # RNN layer with 32 units
    Dense(10, activation='softmax')       # Output layer with 10 units for 10 classes
])

# Step 3: Compile the model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Step 4: Train the model
model.fit(x_train, y_train, epochs=5, batch_size=128, validation_split=0.2)

# Step 5: Evaluate the model on test data
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print('Test Loss: ', test_loss)
print('Test Accuracy: ', test_accuracy)


Epoch 1/5
375/375 [==============================] - 5s 10ms/step - loss: 1.0795 - accuracy: 0.6515 - val_loss: 0.5820 - val_accuracy: 0.8171
Epoch 2/5
375/375 [==============================] - 4s 10ms/step - loss: 0.5004 - accuracy: 0.8484 - val_loss: 0.3999 - val_accuracy: 0.8807
Epoch 3/5
375/375 [==============================] - 4s 9ms/step - loss: 0.3966 - accuracy: 0.8793 - val_loss: 0.3419 - val_accuracy: 0.8994
Epoch 4/5
375/375 [==============================] - 3s 9ms/step - loss: 0.3544 - accuracy: 0.8926 - val_loss: 0.3054 - val_accuracy: 0.9102
Epoch 5/5
313/313 [==============================] - 1s 3ms/step - loss: 0.2999 - accuracy: 0.9102
Test Loss:  0.2998817265033722
Test Accuracy:  0.9101999998092651


## PReLU

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import SimpleRNN, Dense
from tensorflow.keras.utils import to_categorical

# Step 1: Load and preprocess the MNIST dataset
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train.reshape(-1, 28, 28) / 255.0  # Reshape and normalize training data
x_test = x_test.reshape(-1, 28, 28) / 255.0    # Reshape and normalize test data
y_train = to_categorical(y_train, num_classes=10)  # One-hot encode training labels
y_test = to_categorical(y_test, num_classes=10)    # One-hot encode test labels

# Step 2: Define the RNN model
model = Sequential([
    SimpleRNN(32, input_shape=(28, 28), activation = 'PReLU'),  # RNN layer with 32 units
    Dense(10, activation='softmax')       # Output layer with 10 units for 10 classes
])

# Step 3: Compile the model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Step 4: Train the model
model.fit(x_train, y_train, epochs=5, batch_size=128, validation_split=0.2)

# Step 5: Evaluate the model on test data
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print('Test Loss: ', test_loss)
print('Test Accuracy: ', test_accuracy)


Epoch 1/5
375/375 [==============================] - 5s 11ms/step - loss: 1.2793 - accuracy: 0.5543 - val_loss: 0.7745 - val_accuracy: 0.7333
Epoch 2/5
375/375 [==============================] - 11s 30ms/step - loss: 0.6984 - accuracy: 0.7718 - val_loss: 0.5570 - val_accuracy: 0.8149
Epoch 3/5
375/375 [==============================] - 4s 11ms/step - loss: 0.4997 - accuracy: 0.8476 - val_loss: 0.4105 - val_accuracy: 0.8731
Epoch 4/5
375/375 [==============================] - 4s 11ms/step - loss: 0.3955 - accuracy: 0.8832 - val_loss: 0.3246 - val_accuracy: 0.9035
Epoch 5/5
313/313 [==============================] - 1s 3ms/step - loss: 0.3123 - accuracy: 0.9065
Test Loss:  0.31234365701675415
Test Accuracy:  0.906499981880188


## Sigmoid

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import SimpleRNN, Dense
from tensorflow.keras.utils import to_categorical

# Step 1: Load and preprocess the MNIST dataset
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train.reshape(-1, 28, 28) / 255.0  # Reshape and normalize training data
x_test = x_test.reshape(-1, 28, 28) / 255.0    # Reshape and normalize test data
y_train = to_categorical(y_train, num_classes=10)  # One-hot encode training labels
y_test = to_categorical(y_test, num_classes=10)    # One-hot encode test labels

# Step 2: Define the RNN model
model = Sequential([
    SimpleRNN(32, input_shape=(28, 28), activation = 'sigmoid'),  # RNN layer with 32 units
    Dense(10, activation='softmax')       # Output layer with 10 units for 10 classes
])

# Step 3: Compile the model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Step 4: Train the model
model.fit(x_train, y_train, epochs=5, batch_size=128, validation_split=0.2)

# Step 5: Evaluate the model on test data
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print('Test Loss: ', test_loss)
print('Test Accuracy: ', test_accuracy)


Epoch 1/5
375/375 [==============================] - 7s 13ms/step - loss: 2.2841 - accuracy: 0.1558 - val_loss: 2.1618 - val_accuracy: 0.2624
Epoch 2/5
375/375 [==============================] - 4s 10ms/step - loss: 1.9302 - accuracy: 0.2995 - val_loss: 1.7839 - val_accuracy: 0.3292
Epoch 3/5
375/375 [==============================] - 4s 11ms/step - loss: 1.7700 - accuracy: 0.3486 - val_loss: 1.7167 - val_accuracy: 0.3363
Epoch 4/5
375/375 [==============================] - 3s 9ms/step - loss: 1.7297 - accuracy: 0.3510 - val_loss: 1.6888 - val_accuracy: 0.3512
Epoch 5/5
313/313 [==============================] - 1s 3ms/step - loss: 1.7021 - accuracy: 0.3429
Test Loss:  1.702112078666687
Test Accuracy:  0.34290000796318054


## tanh

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import SimpleRNN, Dense
from tensorflow.keras.utils import to_categorical

# Step 1: Load and preprocess the MNIST dataset
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train.reshape(-1, 28, 28) / 255.0  # Reshape and normalize training data
x_test = x_test.reshape(-1, 28, 28) / 255.0    # Reshape and normalize test data
y_train = to_categorical(y_train, num_classes=10)  # One-hot encode training labels
y_test = to_categorical(y_test, num_classes=10)    # One-hot encode test labels

# Step 2: Define the RNN model
model = Sequential([
    SimpleRNN(32, input_shape=(28, 28), activation = 'tanh'),  # RNN layer with 32 units
    Dense(10, activation='softmax')       # Output layer with 10 units for 10 classes
])

# Step 3: Compile the model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Step 4: Train the model
model.fit(x_train, y_train, epochs=5, batch_size=128, validation_split=0.2)

# Step 5: Evaluate the model on test data
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print('Test Loss: ', test_loss)
print('Test Accuracy: ', test_accuracy)


Epoch 1/5
375/375 [==============================] - 5s 10ms/step - loss: 1.3672 - accuracy: 0.5425 - val_loss: 0.8915 - val_accuracy: 0.7138
Epoch 2/5
375/375 [==============================] - 3s 8ms/step - loss: 0.7893 - accuracy: 0.7436 - val_loss: 0.6193 - val_accuracy: 0.8142
Epoch 3/5
375/375 [==============================] - 3s 9ms/step - loss: 0.5989 - accuracy: 0.8195 - val_loss: 0.5093 - val_accuracy: 0.8495
Epoch 4/5
375/375 [==============================] - 4s 11ms/step - loss: 0.4930 - accuracy: 0.8577 - val_loss: 0.4388 - val_accuracy: 0.8742
Epoch 5/5
313/313 [==============================] - 1s 4ms/step - loss: 0.4027 - accuracy: 0.8860
Test Loss:  0.40266937017440796
Test Accuracy:  0.8859999775886536


## softmax

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import SimpleRNN, Dense
from tensorflow.keras.utils import to_categorical

# Step 1: Load and preprocess the MNIST dataset
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train.reshape(-1, 28, 28) / 255.0  # Reshape and normalize training data
x_test = x_test.reshape(-1, 28, 28) / 255.0    # Reshape and normalize test data
y_train = to_categorical(y_train, num_classes=10)  # One-hot encode training labels
y_test = to_categorical(y_test, num_classes=10)    # One-hot encode test labels

# Step 2: Define the RNN model
model = Sequential([
    SimpleRNN(32, input_shape=(28, 28), activation = 'softmax'),  # RNN layer with 32 units
    Dense(10, activation='softmax')       # Output layer with 10 units for 10 classes
])

# Step 3: Compile the model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Step 4: Train the model
model.fit(x_train, y_train, epochs=5, batch_size=128, validation_split=0.2)

# Step 5: Evaluate the model on test data
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print('Test Loss: ', test_loss)
print('Test Accuracy: ', test_accuracy)


Epoch 1/5
375/375 [==============================] - 6s 11ms/step - loss: 2.3012 - accuracy: 0.1116 - val_loss: 2.3001 - val_accuracy: 0.1091
Epoch 2/5
375/375 [==============================] - 4s 10ms/step - loss: 2.2879 - accuracy: 0.1473 - val_loss: 2.2491 - val_accuracy: 0.1783
Epoch 3/5
375/375 [==============================] - 4s 11ms/step - loss: 2.1448 - accuracy: 0.2189 - val_loss: 2.0407 - val_accuracy: 0.2620
Epoch 4/5
375/375 [==============================] - 4s 10ms/step - loss: 1.9964 - accuracy: 0.2316 - val_loss: 1.9483 - val_accuracy: 0.2673
Epoch 5/5
313/313 [==============================] - 1s 5ms/step - loss: 1.9007 - accuracy: 0.3146
Test Loss:  1.9007335901260376
Test Accuracy:  0.31459999084472656


## conclusion

Best activation function: elu, Test Accuracy:  0.9101999998092651

# Best activation function: elu, 8 optimizer

## SGD

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import SimpleRNN, Dense
from tensorflow.keras.utils import to_categorical

# Step 1: Load and preprocess the MNIST dataset
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train.reshape(-1, 28, 28) / 255.0  # Reshape and normalize training data
x_test = x_test.reshape(-1, 28, 28) / 255.0    # Reshape and normalize test data
y_train = to_categorical(y_train, num_classes=10)  # One-hot encode training labels
y_test = to_categorical(y_test, num_classes=10)    # One-hot encode test labels

# Step 2: Define the RNN model
model = Sequential([
    SimpleRNN(32, input_shape=(28, 28), activation = 'elu'),  # RNN layer with 32 units
    Dense(10, activation='softmax')       # Output layer with 10 units for 10 classes
])

# Step 3: Compile the model
model.compile(optimizer='SGD', loss='categorical_crossentropy', metrics=['accuracy'])

# Step 4: Train the model
model.fit(x_train, y_train, epochs=5, batch_size=128, validation_split=0.2)

# Step 5: Evaluate the model on test data
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print('Test Loss: ', test_loss)
print('Test Accuracy: ', test_accuracy)


Epoch 1/5
375/375 [==============================] - 4s 10ms/step - loss: 1.7769 - accuracy: 0.3989 - val_loss: 1.1593 - val_accuracy: 0.5932
Epoch 2/5
375/375 [==============================] - 5s 14ms/step - loss: 0.9900 - accuracy: 0.6543 - val_loss: 0.8508 - val_accuracy: 0.6989
Epoch 3/5
375/375 [==============================] - 4s 11ms/step - loss: 0.7666 - accuracy: 0.7392 - val_loss: 0.8339 - val_accuracy: 0.6955
Epoch 4/5
375/375 [==============================] - 3s 9ms/step - loss: 0.6393 - accuracy: 0.7827 - val_loss: 0.5151 - val_accuracy: 0.8261
Epoch 5/5
313/313 [==============================] - 2s 5ms/step - loss: 0.4723 - accuracy: 0.8429
Test Loss:  0.47225525975227356
Test Accuracy:  0.8428999781608582


## SGD with Momentum

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import SimpleRNN, Dense
from tensorflow.keras.utils import to_categorical

# Step 1: Load and preprocess the MNIST dataset
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train.reshape(-1, 28, 28) / 255.0  # Reshape and normalize training data
x_test = x_test.reshape(-1, 28, 28) / 255.0    # Reshape and normalize test data
y_train = to_categorical(y_train, num_classes=10)  # One-hot encode training labels
y_test = to_categorical(y_test, num_classes=10)    # One-hot encode test labels

# Step 2: Define the RNN model
model = Sequential([
    SimpleRNN(32, input_shape=(28, 28), activation = 'elu'),  # RNN layer with 32 units
    Dense(10, activation='softmax')       # Output layer with 10 units for 10 classes
])

# Step 3: Compile the model
SGDm = tf.keras.optimizers.SGD(learning_rate=0.01, momentum=0.9)
model.compile(optimizer=SGDm, loss='categorical_crossentropy', metrics=['accuracy'])

# Step 4: Train the model
model.fit(x_train, y_train, epochs=5, batch_size=128, validation_split=0.2)

# Step 5: Evaluate the model on test data
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print('Test Loss: ', test_loss)
print('Test Accuracy: ', test_accuracy)


Epoch 1/5
375/375 [==============================] - 5s 10ms/step - loss: 0.9315 - accuracy: 0.6812 - val_loss: 0.5775 - val_accuracy: 0.8242
Epoch 2/5
375/375 [==============================] - 3s 9ms/step - loss: 0.4700 - accuracy: 0.8565 - val_loss: 0.3108 - val_accuracy: 0.9094
Epoch 3/5
375/375 [==============================] - 4s 11ms/step - loss: 0.3639 - accuracy: 0.8935 - val_loss: 0.2960 - val_accuracy: 0.9178
Epoch 4/5
375/375 [==============================] - 3s 8ms/step - loss: 0.3033 - accuracy: 0.9113 - val_loss: 0.2206 - val_accuracy: 0.9347
Epoch 5/5
313/313 [==============================] - 1s 3ms/step - loss: 0.2522 - accuracy: 0.9248
Test Loss:  0.25216323137283325
Test Accuracy:  0.9247999787330627


## RMSProp

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import SimpleRNN, Dense
from tensorflow.keras.utils import to_categorical

# Step 1: Load and preprocess the MNIST dataset
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train.reshape(-1, 28, 28) / 255.0  # Reshape and normalize training data
x_test = x_test.reshape(-1, 28, 28) / 255.0    # Reshape and normalize test data
y_train = to_categorical(y_train, num_classes=10)  # One-hot encode training labels
y_test = to_categorical(y_test, num_classes=10)    # One-hot encode test labels

# Step 2: Define the RNN model
model = Sequential([
    SimpleRNN(32, input_shape=(28, 28), activation = 'elu'),  # RNN layer with 32 units
    Dense(10, activation='softmax')       # Output layer with 10 units for 10 classes
])

# Step 3: Compile the model
model.compile(optimizer='RMSProp', loss='categorical_crossentropy', metrics=['accuracy'])

# Step 4: Train the model
model.fit(x_train, y_train, epochs=5, batch_size=128, validation_split=0.2)

# Step 5: Evaluate the model on test data
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print('Test Loss: ', test_loss)
print('Test Accuracy: ', test_accuracy)


Epoch 1/5
375/375 [==============================] - 6s 13ms/step - loss: 1.0421 - accuracy: 0.6525 - val_loss: 0.6515 - val_accuracy: 0.7911
Epoch 2/5
375/375 [==============================] - 5s 14ms/step - loss: 0.6086 - accuracy: 0.8021 - val_loss: 0.4660 - val_accuracy: 0.8506
Epoch 3/5
375/375 [==============================] - 4s 11ms/step - loss: 0.4652 - accuracy: 0.8559 - val_loss: 0.3916 - val_accuracy: 0.8851
Epoch 4/5
375/375 [==============================] - 4s 11ms/step - loss: 0.3806 - accuracy: 0.8857 - val_loss: 0.3354 - val_accuracy: 0.9032
Epoch 5/5
313/313 [==============================] - 1s 3ms/step - loss: 0.2704 - accuracy: 0.9214
Test Loss:  0.27042049169540405
Test Accuracy:  0.9214000105857849


## adam

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import SimpleRNN, Dense
from tensorflow.keras.utils import to_categorical

# Step 1: Load and preprocess the MNIST dataset
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train.reshape(-1, 28, 28) / 255.0  # Reshape and normalize training data
x_test = x_test.reshape(-1, 28, 28) / 255.0    # Reshape and normalize test data
y_train = to_categorical(y_train, num_classes=10)  # One-hot encode training labels
y_test = to_categorical(y_test, num_classes=10)    # One-hot encode test labels

# Step 2: Define the RNN model
model = Sequential([
    SimpleRNN(32, input_shape=(28, 28), activation = 'elu'),  # RNN layer with 32 units
    Dense(10, activation='softmax')       # Output layer with 10 units for 10 classes
])

# Step 3: Compile the model
model.compile(optimizer='Adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Step 4: Train the model
model.fit(x_train, y_train, epochs=5, batch_size=128, validation_split=0.2)

# Step 5: Evaluate the model on test data
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print('Test Loss: ', test_loss)
print('Test Accuracy: ', test_accuracy)


Epoch 1/5
375/375 [==============================] - 5s 10ms/step - loss: 0.9820 - accuracy: 0.6658 - val_loss: 0.4918 - val_accuracy: 0.8448
Epoch 2/5
375/375 [==============================] - 3s 9ms/step - loss: 0.4649 - accuracy: 0.8566 - val_loss: 0.3971 - val_accuracy: 0.8764
Epoch 3/5
375/375 [==============================] - 3s 9ms/step - loss: 0.3810 - accuracy: 0.8830 - val_loss: 0.3601 - val_accuracy: 0.8898
Epoch 4/5
375/375 [==============================] - 4s 11ms/step - loss: 0.3335 - accuracy: 0.8982 - val_loss: 0.2919 - val_accuracy: 0.9115
Epoch 5/5
313/313 [==============================] - 1s 3ms/step - loss: 0.2804 - accuracy: 0.9132
Test Loss:  0.28041738271713257
Test Accuracy:  0.9132000207901001


## adamax

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import SimpleRNN, Dense
from tensorflow.keras.utils import to_categorical

# Step 1: Load and preprocess the MNIST dataset
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train.reshape(-1, 28, 28) / 255.0  # Reshape and normalize training data
x_test = x_test.reshape(-1, 28, 28) / 255.0    # Reshape and normalize test data
y_train = to_categorical(y_train, num_classes=10)  # One-hot encode training labels
y_test = to_categorical(y_test, num_classes=10)    # One-hot encode test labels

# Step 2: Define the RNN model
model = Sequential([
    SimpleRNN(32, input_shape=(28, 28), activation = 'elu'),  # RNN layer with 32 units
    Dense(10, activation='softmax')       # Output layer with 10 units for 10 classes
])

# Step 3: Compile the model
model.compile(optimizer='Adamax', loss='categorical_crossentropy', metrics=['accuracy'])

# Step 4: Train the model
model.fit(x_train, y_train, epochs=5, batch_size=128, validation_split=0.2)

# Step 5: Evaluate the model on test data
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print('Test Loss: ', test_loss)
print('Test Accuracy: ', test_accuracy)


Epoch 1/5
375/375 [==============================] - 5s 10ms/step - loss: 1.3459 - accuracy: 0.5323 - val_loss: 0.7872 - val_accuracy: 0.7459
Epoch 2/5
375/375 [==============================] - 6s 15ms/step - loss: 0.6801 - accuracy: 0.7771 - val_loss: 0.5402 - val_accuracy: 0.8263
Epoch 3/5
375/375 [==============================] - 3s 9ms/step - loss: 0.5355 - accuracy: 0.8261 - val_loss: 0.4592 - val_accuracy: 0.8548
Epoch 4/5
375/375 [==============================] - 5s 14ms/step - loss: 0.4631 - accuracy: 0.8529 - val_loss: 0.4128 - val_accuracy: 0.8691
Epoch 5/5
313/313 [==============================] - 1s 3ms/step - loss: 0.3870 - accuracy: 0.8800
Test Loss:  0.38695552945137024
Test Accuracy:  0.8799999952316284


## nadam

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import SimpleRNN, Dense
from tensorflow.keras.utils import to_categorical

# Step 1: Load and preprocess the MNIST dataset
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train.reshape(-1, 28, 28) / 255.0  # Reshape and normalize training data
x_test = x_test.reshape(-1, 28, 28) / 255.0    # Reshape and normalize test data
y_train = to_categorical(y_train, num_classes=10)  # One-hot encode training labels
y_test = to_categorical(y_test, num_classes=10)    # One-hot encode test labels

# Step 2: Define the RNN model
model = Sequential([
    SimpleRNN(32, input_shape=(28, 28), activation = 'elu'),  # RNN layer with 32 units
    Dense(10, activation='softmax')       # Output layer with 10 units for 10 classes
])

# Step 3: Compile the model
model.compile(optimizer='Nadam', loss='categorical_crossentropy', metrics=['accuracy'])

# Step 4: Train the model
model.fit(x_train, y_train, epochs=5, batch_size=128, validation_split=0.2)

# Step 5: Evaluate the model on test data
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print('Test Loss: ', test_loss)
print('Test Accuracy: ', test_accuracy)


Epoch 1/5
375/375 [==============================] - 6s 11ms/step - loss: 0.9771 - accuracy: 0.6699 - val_loss: 0.4766 - val_accuracy: 0.8523
Epoch 2/5
375/375 [==============================] - 3s 9ms/step - loss: 0.4589 - accuracy: 0.8601 - val_loss: 0.3738 - val_accuracy: 0.8857
Epoch 3/5
375/375 [==============================] - 4s 10ms/step - loss: 0.3767 - accuracy: 0.8873 - val_loss: 0.3161 - val_accuracy: 0.9058
Epoch 4/5
375/375 [==============================] - 4s 10ms/step - loss: 0.3273 - accuracy: 0.9017 - val_loss: 0.2815 - val_accuracy: 0.9139
Epoch 5/5
313/313 [==============================] - 1s 3ms/step - loss: 0.2693 - accuracy: 0.9197
Test Loss:  0.2693077027797699
Test Accuracy:  0.919700026512146


## AdaGrad

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import SimpleRNN, Dense
from tensorflow.keras.utils import to_categorical

# Step 1: Load and preprocess the MNIST dataset
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train.reshape(-1, 28, 28) / 255.0  # Reshape and normalize training data
x_test = x_test.reshape(-1, 28, 28) / 255.0    # Reshape and normalize test data
y_train = to_categorical(y_train, num_classes=10)  # One-hot encode training labels
y_test = to_categorical(y_test, num_classes=10)    # One-hot encode test labels

# Step 2: Define the RNN model
model = Sequential([
    SimpleRNN(32, input_shape=(28, 28), activation = 'elu'),  # RNN layer with 32 units
    Dense(10, activation='softmax')       # Output layer with 10 units for 10 classes
])

# Step 3: Compile the model
model.compile(optimizer='AdaGrad', loss='categorical_crossentropy', metrics=['accuracy'])

# Step 4: Train the model
model.fit(x_train, y_train, epochs=5, batch_size=128, validation_split=0.2)

# Step 5: Evaluate the model on test data
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print('Test Loss: ', test_loss)
print('Test Accuracy: ', test_accuracy)


Epoch 1/5
375/375 [==============================] - 5s 10ms/step - loss: 2.2252 - accuracy: 0.2140 - val_loss: 2.0808 - val_accuracy: 0.2888
Epoch 2/5
375/375 [==============================] - 3s 8ms/step - loss: 1.9569 - accuracy: 0.3225 - val_loss: 1.8016 - val_accuracy: 0.3997
Epoch 3/5
375/375 [==============================] - 3s 9ms/step - loss: 1.7061 - accuracy: 0.4341 - val_loss: 1.5805 - val_accuracy: 0.4880
Epoch 4/5
375/375 [==============================] - 4s 11ms/step - loss: 1.5309 - accuracy: 0.4966 - val_loss: 1.4366 - val_accuracy: 0.5328
Epoch 5/5
313/313 [==============================] - 1s 3ms/step - loss: 1.3659 - accuracy: 0.5452
Test Loss:  1.3658630847930908
Test Accuracy:  0.545199990272522


## AdaDelta

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import SimpleRNN, Dense
from tensorflow.keras.utils import to_categorical

# Step 1: Load and preprocess the MNIST dataset
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train.reshape(-1, 28, 28) / 255.0  # Reshape and normalize training data
x_test = x_test.reshape(-1, 28, 28) / 255.0    # Reshape and normalize test data
y_train = to_categorical(y_train, num_classes=10)  # One-hot encode training labels
y_test = to_categorical(y_test, num_classes=10)    # One-hot encode test labels

# Step 2: Define the RNN model
model = Sequential([
    SimpleRNN(32, input_shape=(28, 28), activation = 'elu'),  # RNN layer with 32 units
    Dense(10, activation='softmax')       # Output layer with 10 units for 10 classes
])

# Step 3: Compile the model
model.compile(optimizer='AdaDelta', loss='categorical_crossentropy', metrics=['accuracy'])

# Step 4: Train the model
model.fit(x_train, y_train, epochs=5, batch_size=128, validation_split=0.2)

# Step 5: Evaluate the model on test data
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print('Test Loss: ', test_loss)
print('Test Accuracy: ', test_accuracy)


Epoch 1/5
375/375 [==============================] - 5s 10ms/step - loss: 2.5800 - accuracy: 0.1105 - val_loss: 2.5310 - val_accuracy: 0.1164
Epoch 2/5
375/375 [==============================] - 3s 9ms/step - loss: 2.5120 - accuracy: 0.1174 - val_loss: 2.4694 - val_accuracy: 0.1245
Epoch 3/5
375/375 [==============================] - 4s 11ms/step - loss: 2.4556 - accuracy: 0.1241 - val_loss: 2.4189 - val_accuracy: 0.1316
Epoch 4/5
375/375 [==============================] - 3s 8ms/step - loss: 2.4092 - accuracy: 0.1330 - val_loss: 2.3777 - val_accuracy: 0.1379
Epoch 5/5
313/313 [==============================] - 2s 5ms/step - loss: 2.3562 - accuracy: 0.1462
Test Loss:  2.3562207221984863
Test Accuracy:  0.1462000012397766


## conclusion
Best optimizer: SGD with Momentum. Test Accuracy:  0.9247999787330627

#Experimenting Best Output Activation Function


## output relu

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import SimpleRNN, Dense
from tensorflow.keras.utils import to_categorical

# Step 1: Load and preprocess the MNIST dataset
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train.reshape(-1, 28, 28) / 255.0  # Reshape and normalize training data
x_test = x_test.reshape(-1, 28, 28) / 255.0    # Reshape and normalize test data
y_train = to_categorical(y_train, num_classes=10)  # One-hot encode training labels
y_test = to_categorical(y_test, num_classes=10)    # One-hot encode test labels

# Step 2: Define the RNN model
model = Sequential([
    SimpleRNN(32, input_shape=(28, 28), activation = 'elu'),  # RNN layer with 32 units
    Dense(10, activation='relu')       # Output layer with 10 units for 10 classes
])

# Step 3: Compile the model
SGDm = tf.keras.optimizers.SGD(learning_rate=0.01, momentum=0.9)
model.compile(optimizer=SGDm, loss='categorical_crossentropy', metrics=['accuracy'])

# Step 4: Train the model
model.fit(x_train, y_train, epochs=5, batch_size=128, validation_split=0.2)

# Step 5: Evaluate the model on test data
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print('Test Loss: ', test_loss)
print('Test Accuracy: ', test_accuracy)


Epoch 1/5
375/375 [==============================] - 7s 17ms/step - loss: 2.5163 - accuracy: 0.1616 - val_loss: 2.2710 - val_accuracy: 0.1674
Epoch 2/5
375/375 [==============================] - 4s 10ms/step - loss: nan - accuracy: 0.1189 - val_loss: nan - val_accuracy: 0.0995
Epoch 3/5
375/375 [==============================] - 3s 9ms/step - loss: nan - accuracy: 0.0985 - val_loss: nan - val_accuracy: 0.0995
Epoch 4/5
375/375 [==============================] - 6s 15ms/step - loss: nan - accuracy: 0.0985 - val_loss: nan - val_accuracy: 0.0995
Epoch 5/5
313/313 [==============================] - 2s 5ms/step - loss: nan - accuracy: 0.0980
Test Loss:  nan
Test Accuracy:  0.09799999743700027


## output LeakyReLU

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import SimpleRNN, Dense
from tensorflow.keras.utils import to_categorical

# Step 1: Load and preprocess the MNIST dataset
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train.reshape(-1, 28, 28) / 255.0  # Reshape and normalize training data
x_test = x_test.reshape(-1, 28, 28) / 255.0    # Reshape and normalize test data
y_train = to_categorical(y_train, num_classes=10)  # One-hot encode training labels
y_test = to_categorical(y_test, num_classes=10)    # One-hot encode test labels

# Step 2: Define the RNN model
model = Sequential([
    SimpleRNN(32, input_shape=(28, 28), activation = 'elu'),  # RNN layer with 32 units
    Dense(10, activation='LeakyReLU')       # Output layer with 10 units for 10 classes
])

# Step 3: Compile the model
SGDm = tf.keras.optimizers.SGD(learning_rate=0.01, momentum=0.9)
model.compile(optimizer=SGDm, loss='categorical_crossentropy', metrics=['accuracy'])

# Step 4: Train the model
model.fit(x_train, y_train, epochs=5, batch_size=128, validation_split=0.2)

# Step 5: Evaluate the model on test data
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print('Test Loss: ', test_loss)
print('Test Accuracy: ', test_accuracy)


Epoch 1/5
375/375 [==============================] - 7s 13ms/step - loss: 6.2256 - accuracy: 0.1001 - val_loss: 4.9156 - val_accuracy: 0.1060
Epoch 2/5
375/375 [==============================] - 3s 9ms/step - loss: 7.5089 - accuracy: 0.0972 - val_loss: 9.7635 - val_accuracy: 0.0914
Epoch 3/5
375/375 [==============================] - 6s 15ms/step - loss: 9.7971 - accuracy: 0.0901 - val_loss: 9.7635 - val_accuracy: 0.0914
Epoch 4/5
375/375 [==============================] - 4s 11ms/step - loss: 9.8740 - accuracy: 0.0901 - val_loss: 9.7635 - val_accuracy: 0.0914
Epoch 5/5
313/313 [==============================] - 1s 3ms/step - loss: 9.9029 - accuracy: 0.0892
Test Loss:  9.902948379516602
Test Accuracy:  0.08919999748468399


## output elu

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import SimpleRNN, Dense
from tensorflow.keras.utils import to_categorical

# Step 1: Load and preprocess the MNIST dataset
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train.reshape(-1, 28, 28) / 255.0  # Reshape and normalize training data
x_test = x_test.reshape(-1, 28, 28) / 255.0    # Reshape and normalize test data
y_train = to_categorical(y_train, num_classes=10)  # One-hot encode training labels
y_test = to_categorical(y_test, num_classes=10)    # One-hot encode test labels

# Step 2: Define the RNN model
model = Sequential([
    SimpleRNN(32, input_shape=(28, 28), activation = 'elu'),  # RNN layer with 32 units
    Dense(10, activation='elu')       # Output layer with 10 units for 10 classes
])

# Step 3: Compile the model
SGDm = tf.keras.optimizers.SGD(learning_rate=0.01, momentum=0.9)
model.compile(optimizer=SGDm, loss='categorical_crossentropy', metrics=['accuracy'])

# Step 4: Train the model
model.fit(x_train, y_train, epochs=5, batch_size=128, validation_split=0.2)

# Step 5: Evaluate the model on test data
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print('Test Loss: ', test_loss)
print('Test Accuracy: ', test_accuracy)


Epoch 1/5
375/375 [==============================] - 5s 11ms/step - loss: nan - accuracy: 0.0975 - val_loss: nan - val_accuracy: 0.0995
Epoch 2/5
375/375 [==============================] - 3s 9ms/step - loss: nan - accuracy: 0.0985 - val_loss: nan - val_accuracy: 0.0995
Epoch 3/5
375/375 [==============================] - 3s 9ms/step - loss: nan - accuracy: 0.0985 - val_loss: nan - val_accuracy: 0.0995
Epoch 4/5
375/375 [==============================] - 3s 9ms/step - loss: nan - accuracy: 0.0985 - val_loss: nan - val_accuracy: 0.0995
Epoch 5/5
313/313 [==============================] - 1s 3ms/step - loss: nan - accuracy: 0.0980
Test Loss:  nan
Test Accuracy:  0.09799999743700027


## output PReLU

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import SimpleRNN, Dense
from tensorflow.keras.utils import to_categorical

# Step 1: Load and preprocess the MNIST dataset
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train.reshape(-1, 28, 28) / 255.0  # Reshape and normalize training data
x_test = x_test.reshape(-1, 28, 28) / 255.0    # Reshape and normalize test data
y_train = to_categorical(y_train, num_classes=10)  # One-hot encode training labels
y_test = to_categorical(y_test, num_classes=10)    # One-hot encode test labels

# Step 2: Define the RNN model
model = Sequential([
    SimpleRNN(32, input_shape=(28, 28), activation = 'elu'),  # RNN layer with 32 units
    Dense(10, activation='PReLU')       # Output layer with 10 units for 10 classes
])

# Step 3: Compile the model
SGDm = tf.keras.optimizers.SGD(learning_rate=0.01, momentum=0.9)
model.compile(optimizer=SGDm, loss='categorical_crossentropy', metrics=['accuracy'])

# Step 4: Train the model
model.fit(x_train, y_train, epochs=5, batch_size=128, validation_split=0.2)

# Step 5: Evaluate the model on test data
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print('Test Loss: ', test_loss)
print('Test Accuracy: ', test_accuracy)


Epoch 1/5
375/375 [==============================] - 5s 10ms/step - loss: 3.2526 - accuracy: 0.0971 - val_loss: 2.9254 - val_accuracy: 0.0998
Epoch 2/5
375/375 [==============================] - 4s 10ms/step - loss: nan - accuracy: 0.0977 - val_loss: nan - val_accuracy: 0.0995
Epoch 3/5
375/375 [==============================] - 3s 8ms/step - loss: nan - accuracy: 0.0985 - val_loss: nan - val_accuracy: 0.0995
Epoch 4/5
375/375 [==============================] - 3s 9ms/step - loss: nan - accuracy: 0.0985 - val_loss: nan - val_accuracy: 0.0995
Epoch 5/5
313/313 [==============================] - 1s 3ms/step - loss: nan - accuracy: 0.0980
Test Loss:  nan
Test Accuracy:  0.09799999743700027


## output sigmoid

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import SimpleRNN, Dense
from tensorflow.keras.utils import to_categorical

# Step 1: Load and preprocess the MNIST dataset
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train.reshape(-1, 28, 28) / 255.0  # Reshape and normalize training data
x_test = x_test.reshape(-1, 28, 28) / 255.0    # Reshape and normalize test data
y_train = to_categorical(y_train, num_classes=10)  # One-hot encode training labels
y_test = to_categorical(y_test, num_classes=10)    # One-hot encode test labels

# Step 2: Define the RNN model
model = Sequential([
    SimpleRNN(32, input_shape=(28, 28), activation = 'elu'),  # RNN layer with 32 units
    Dense(10, activation='sigmoid')       # Output layer with 10 units for 10 classes
])

# Step 3: Compile the model
SGDm = tf.keras.optimizers.SGD(learning_rate=0.01, momentum=0.9)
model.compile(optimizer=SGDm, loss='categorical_crossentropy', metrics=['accuracy'])

# Step 4: Train the model
model.fit(x_train, y_train, epochs=5, batch_size=128, validation_split=0.2)

# Step 5: Evaluate the model on test data
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print('Test Loss: ', test_loss)
print('Test Accuracy: ', test_accuracy)


Epoch 1/5
375/375 [==============================] - 5s 10ms/step - loss: 0.9174 - accuracy: 0.6975 - val_loss: 0.4460 - val_accuracy: 0.8697
Epoch 2/5
375/375 [==============================] - 4s 11ms/step - loss: 0.4988 - accuracy: 0.8479 - val_loss: 0.4447 - val_accuracy: 0.8721
Epoch 3/5
375/375 [==============================] - 3s 9ms/step - loss: 0.4230 - accuracy: 0.8728 - val_loss: 0.3291 - val_accuracy: 0.9057
Epoch 4/5
375/375 [==============================] - 3s 9ms/step - loss: 0.3574 - accuracy: 0.8931 - val_loss: 0.2979 - val_accuracy: 0.9112
Epoch 5/5
313/313 [==============================] - 1s 3ms/step - loss: 0.2627 - accuracy: 0.9242
Test Loss:  0.2627376616001129
Test Accuracy:  0.9241999983787537


## output tanh

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import SimpleRNN, Dense
from tensorflow.keras.utils import to_categorical

# Step 1: Load and preprocess the MNIST dataset
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train.reshape(-1, 28, 28) / 255.0  # Reshape and normalize training data
x_test = x_test.reshape(-1, 28, 28) / 255.0    # Reshape and normalize test data
y_train = to_categorical(y_train, num_classes=10)  # One-hot encode training labels
y_test = to_categorical(y_test, num_classes=10)    # One-hot encode test labels

# Step 2: Define the RNN model
model = Sequential([
    SimpleRNN(32, input_shape=(28, 28), activation = 'elu'),  # RNN layer with 32 units
    Dense(10, activation='tanh')       # Output layer with 10 units for 10 classes
])

# Step 3: Compile the model
SGDm = tf.keras.optimizers.SGD(learning_rate=0.01, momentum=0.9)
model.compile(optimizer=SGDm, loss='categorical_crossentropy', metrics=['accuracy'])

# Step 4: Train the model
model.fit(x_train, y_train, epochs=5, batch_size=128, validation_split=0.2)

# Step 5: Evaluate the model on test data
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print('Test Loss: ', test_loss)
print('Test Accuracy: ', test_accuracy)


Epoch 1/5
375/375 [==============================] - 7s 16ms/step - loss: nan - accuracy: 0.0985 - val_loss: nan - val_accuracy: 0.0995
Epoch 2/5
375/375 [==============================] - 3s 8ms/step - loss: nan - accuracy: 0.0985 - val_loss: nan - val_accuracy: 0.0995
Epoch 3/5
375/375 [==============================] - 3s 9ms/step - loss: nan - accuracy: 0.0985 - val_loss: nan - val_accuracy: 0.0995
Epoch 4/5
375/375 [==============================] - 4s 11ms/step - loss: nan - accuracy: 0.0985 - val_loss: nan - val_accuracy: 0.0995
Epoch 5/5
313/313 [==============================] - 1s 4ms/step - loss: nan - accuracy: 0.0980
Test Loss:  nan
Test Accuracy:  0.09799999743700027


## output softmax

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import SimpleRNN, Dense
from tensorflow.keras.utils import to_categorical

# Step 1: Load and preprocess the MNIST dataset
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train.reshape(-1, 28, 28) / 255.0  # Reshape and normalize training data
x_test = x_test.reshape(-1, 28, 28) / 255.0    # Reshape and normalize test data
y_train = to_categorical(y_train, num_classes=10)  # One-hot encode training labels
y_test = to_categorical(y_test, num_classes=10)    # One-hot encode test labels

# Step 2: Define the RNN model
model = Sequential([
    SimpleRNN(32, input_shape=(28, 28), activation = 'elu'),  # RNN layer with 32 units
    Dense(10, activation='softmax')       # Output layer with 10 units for 10 classes
])

# Step 3: Compile the model
SGDm = tf.keras.optimizers.SGD(learning_rate=0.01, momentum=0.9)
model.compile(optimizer=SGDm, loss='categorical_crossentropy', metrics=['accuracy'])

# Step 4: Train the model
model.fit(x_train, y_train, epochs=5, batch_size=128, validation_split=0.2)

# Step 5: Evaluate the model on test data
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print('Test Loss: ', test_loss)
print('Test Accuracy: ', test_accuracy)


Epoch 1/5
375/375 [==============================] - 7s 13ms/step - loss: 0.9872 - accuracy: 0.6695 - val_loss: 0.4861 - val_accuracy: 0.8523
Epoch 2/5
375/375 [==============================] - 4s 9ms/step - loss: 0.4803 - accuracy: 0.8543 - val_loss: 0.4455 - val_accuracy: 0.8660
Epoch 3/5
375/375 [==============================] - 4s 10ms/step - loss: 0.3771 - accuracy: 0.8885 - val_loss: 0.3034 - val_accuracy: 0.9112
Epoch 4/5
375/375 [==============================] - 4s 12ms/step - loss: 0.3461 - accuracy: 0.8989 - val_loss: 0.2799 - val_accuracy: 0.9179
Epoch 5/5
313/313 [==============================] - 1s 4ms/step - loss: 0.3836 - accuracy: 0.8864
Test Loss:  0.3836445212364197
Test Accuracy:  0.8863999843597412


## conclusion
Best output activation function: sigmoid. Test Accuracy:  0.9241999983787537